File: learning/02_GettingStarted/py/src/mm_count.py
import sys

from pyspark.sql import SparkSession


def main():
    print('Python version:', sys.version)
    if len(sys.argv) != 2:
        print('Usage: mm_count [file]', file=sys.stderr)
        sys.exit(-1)

    # Build a SparkSession (can only be one sess per JVM)
    spark = SparkSession.builder.appName('PythonMMCount').getOrCreate()
    mm_file = sys.argv[1]
    mm_df = (spark.read.format('csv')
             .option('header', 'true')
             .option('inferSchema', 'true')
             .load(mm_file))
    count_df = (mm_df.select('State', 'Color', 'Count')
                .groupBy('State', 'Color')
                .sum('Count')
                .orderBy('sum(Count)', ascending=False))
    count_df.show(n=60, truncate=False)
    print(f'Total Rows: {count_df.count()}')

    # Show data just for a single state
    ca_count_df = (mm_df.select('State', 'Color', 'Count')
                   .where(mm_df.State == 'CA')
                   .groupBy('State', 'Color')
                   .sum('Count')
                   .orderBy('sum(Count)', ascending=False))
    ca_count_df.show(n=10, truncate=False)
    spark.stop()
    
    
if __name__ == '__main__':
    main()

File: learning/02_GettingStarted/sc/src/main/scala/chapter2/MMCount.scala
package main.scala.chapter2

import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._


/**
 * Usage: MMCount [mm_dataset_file]
 */
object MMCount {
  def main(args: Array[String]) {
    val spark = SparkSession.builder.appName("MMCount").getOrCreate()
    if (args.length < 1) {
      println("Usage: MMCount [mm_dataset_file]")
      sys.exit(1)
    }
    val mmFile = args(0)
    val mmDF = spark
      .read.format("csv")
      .option("header", "true")
      .option("inferSchema", "true")
      .load(mmFile)
    val  countMMDF = mmDF
      .select("State", "Color", "Count")
      .groupBy("State", "Color")
      .sum("Count")
      .orderBy(desc("sum(Count)"))
    countMMDF.show(60)
    println(s"Total Rows: ${countMMDF.count()}")
    println()
    val caCounts = mmDF
      .select("*")
      .where(col("State") === "CA")
      .groupBy("State", "Color")
      .sum("Count")
      .orderBy(desc("sum(Count)"))
    caCounts.show(10)
    spark.stop()
  }
}
